index.js

```diff
@@ -73,45 +73,44 @@ function processEvent (event, context, callback) {
       .update(opts, decoded)
       .then(() => {
         log.info('Wrote ' + decoded.length + ' successfully')
-        callback(null, 'Done!')
+        resolve('Done!')
       })
       .catch((error) => {
         log.error('processEvent: error: ' + error)
         log.trace(error)
-        callback(error)
+        reject(error)
       })
   }).catch((error) => {
     log.error(`processEvent: Error fetching schema (${bibItemOrHolding})`)
     log.trace(error)
-    callback(error)
+    reject(error)
   })
 }
 
-exports.handler = (event, context, callback) => {
-  // Use of libpq seems to cause something to hang out in the event loop, so
-  // tell the lambda engine to just kill the process when callback called:
-  // TODO see if we can remove this now that we're no longer using libpq
-  context.callbackWaitsForEmptyEventLoop = false
-
+exports.handler = async (event) => {
   if (!log) {
     // Set log level (default 'info')
     log = require('loglevel')
-    log.setLevel(process.env['LOGLEVEL'] || 'info')
+    log.setLevel(process.env.LOGLEVEL || 'info')
     log.info('Loading Lambda function')
   }
 
-  if (decryptedDbConnectionString) {
-    db.setConnectionString(decryptedDbConnectionString)
-    processEvent(event, context, callback)
-  } else {
-    // Decrypt code should run once and variables stored outside of the function
-    // handler so that these are decrypted once per container
-    kmsHelper.decryptDbCreds().then((decryptedDbConnectionString) => {
+  const eventPromise = new Promise((resolve, reject) => {
+    if (decryptedDbConnectionString) {
       db.setConnectionString(decryptedDbConnectionString)
-      processEvent(event, context, callback)
-    }).catch((err) => {
-      log.error('Decrypt error:', err)
-      return callback(err)
-    })
-  }
+      processEvent(event, resolve, reject)
+    } else {
+      // Decrypt code should run once and variables stored outside of the function
+      // handler so that these are decrypted once per container
+      kmsHelper.decryptDbCreds().then((decryptedDbConnectionString) => {
+        db.setConnectionString(decryptedDbConnectionString)
+        processEvent(event, resolve, reject)
+      }).catch((err) => {
+        log.error('Decrypt error:', err)
+        reject(err)
+      })
+    }
+  })
+
+  return eventPromise
```

Comment:
This is the most significant change, swapping out a `callback` for use a Promise. Again I'm not 100% sure about this, but [AWS documentation](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html) and the comment here suggest that the event loop is causing some issues. According to the documentation using an `async` handler should change this and hopefully improve this functionality.

Reply:
yeah, tbh I'm far from certain that it will have an impact, but it was enough to switch this over so it seemed to make sense to give it a try.

There is *something* going on with large batches of items though, I just can't figure out what exactly. It hangs for 30 seconds and doesn't output a few log lines that just report the status of the function call towards the end. I guess its unlikely to be this, but I can't see what it could be.

I'll plan on merging this to QA early next week with a plan in place to roll everything back if it turns out to be hugely problematic